avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/StatementInfo.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.avatica.jdbc;
+
+import com.google.common.base.Preconditions;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Objects;
+
+/**
+ *  All we know about a statement. Encapsulates a {@link ResultSet}.
+ */
+public class StatementInfo {
+  private static final String NEXT_METHOD_NAME = "next";
+
+  final Statement statement; // sometimes a PreparedStatement
+  private ResultSet resultSet; // a Proxy, not the real ResultSet.
+  private long position = 0;
+
+  /**
+   * Support a Proxy against ResultSet in which <code>position</code> is incremented after
+   * every call to <code>next()</code>.
+   */
+  private class ResultSetInvocationHandler implements InvocationHandler {
```

Comment:
Actually, it's not quite clear to me how I can get the same result using MethodHandles. I used the Proxy to avoid creating a special method on StatementInfo that needed to be called instead of directly on the ResultSet. I can just do away with that too. It's not a big sticking point to me.